xbmc/utils/GUIUtils.cpp

```diff
@@ -0,0 +1,363 @@
+    CServiceBroker::GetPlaylistPlayer().Reset();
+    CServiceBroker::GetPlaylistPlayer().Add(PLAYLIST_MUSIC, queuedItems);
+    CServiceBroker::GetPlaylistPlayer().SetCurrentPlaylist(PLAYLIST_MUSIC);
+    CServiceBroker::GetPlaylistPlayer().Play();
+  }
+  else // song, so just play it
+  {
+    CFileItem item(itemToPlay->GetItemToPlay());
+    PlaySong(item);
+  }
+  m_musicDatabase.Close();
+  return true;
+}
+
+bool CGUIUtils::AddItemToPlayList(const CFileItemPtr& pItem, CFileItemList& queuedItems)
+{
+  int jobid = CJobManager::GetInstance().AddJob(new CAddItemToPlaylistJob(pItem, queuedItems),
+                                                nullptr, CJob::PRIORITY_NORMAL);
+
+  if (!CGUIDialogBusy::WaitOnEvent(m_utilsJobEvent, 500)) // 500ms before busy dialog appears
+  {
+    // Cancel job if its still waiting in queue (unlikely) or still running
+    CJobManager::GetInstance().CancelJob(jobid);
+    return false;
+  }
+  return true;
+}
+
+void CGUIUtils::PlaySong(CFileItem& item)
+{
+  if (item.HasMusicInfoTag())
+    CServiceBroker::GetPlaylistPlayer().Play(std::make_shared<CFileItem>(item), "");
+}
+
+void CGUIUtils::AddToPlayList(const CFileItemPtr& pItem, CFileItemList& queuedItems)
+{
+  if (!pItem->CanQueue() || pItem->IsRAR() || pItem->IsZIP() || pItem->IsParentFolder())
+    return;
+
+  // fast lookup is needed here
+  queuedItems.SetFastLookup(true);
+
+  if (pItem->IsMusicDb() && pItem->m_bIsFolder && !pItem->IsParentFolder())
+  { // we have a music database folder, just grab the "all" item underneath it
+    XFILE::CMusicDatabaseDirectory dir;
+
+    if (!dir.ContainsSongs(pItem->GetPath()))
+    { // grab the ALL item in this category
+      // Genres will still require 2 lookups, and queuing the entire Genre folder
+      // will require 3 lookups (genre, artist, album)
+      CMusicDbUrl musicUrl;
+      if (musicUrl.FromString(pItem->GetPath()))
+      {
+        musicUrl.AppendPath("-1/");
+        CFileItemPtr item(new CFileItem(musicUrl.ToString(), true));
+        item->SetCanQueue(true); // workaround for CanQueue() check above
+        AddToPlayList(item, queuedItems);
+      }
+      return;
+    }
+  }
+  if (pItem->m_bIsFolder)
+  {
+    // Check if we add a locked share
+    if (pItem->m_bIsShareOrDrive)
+    {
+      CFileItem item = *pItem;
+      if (!g_passwordManager.IsItemUnlocked(&item, "music"))
+        return;
+    }
+
+    // recursive
+    CFileItemList items;
+    GetDirectory(pItem->GetPath(), items);
+    FormatAndSort(items);
+    for (int i = 0; i < items.Size(); ++i)
+      AddToPlayList(items[i], queuedItems);
+  }
+  else
+  {
+    if (pItem->IsPlayList())
+    {
+      std::unique_ptr<PLAYLIST::CPlayList> pPlayList(PLAYLIST::CPlayListFactory::Create(*pItem));
+      if (pPlayList)
+      {
+        // load it
+        if (!pPlayList->Load(pItem->GetPath()))
+        {
+          HELPERS::ShowOKDialogText(CVariant{6}, CVariant{477});
```

Comment:
can this be further broken out? Doesn't make sense to block the execution depending on user input here.

Reply:
Does it not?  That will only fire if the playlist fails to load for some reason (corrupted file or whatever), in which case we notify the user and abort trying to play it.  Actually though, looking at the strings, `6` appears to be `Kodi GIT`.  This bit was just c/p but I think from looking at that that 6 is not correct.  I've never had a failure here though to see the actual dialog box heading!

Thanks for the review btw :)

Reply:
IMO it does not because you can execute this from multiple code paths. Only in those initiated from GUI/GUIWindows it makes sense to show the dialog. If this is executed from json it doesn't make much sense specially because you could have no GUI to wait for user input. Ideally this should be even more broken down and return some status code so that the callee can then take the appropriate action:
- GUI show dialog and wait for user input
- No-GUI log error or something else

Reply:
That's true....But currently, this is a direct lift from https://github.com/xbmc/xbmc/blob/master/xbmc/music/windows/GUIWindowMusicBase.cpp#L493-L505 As such, it wasn't callable from JSON or anywhere else.  In part, this relates to what I said in the PR description about having just one code path for doing foo.  ATM, there are different methods in different places that are called to play stuff.  The end result is the same (item gets played) but the paths are different.  Ultimately, this needs resolving so that JSON, GUI or whatever all can use the same routine/code path to play/queue a music item.
This though is for the future and not for this particular PR.  As it stands, nothing else is calling into that routine than was before, it's merely moved to remove the dependency of having an instance of that gui window.

Reply:
> HELPERS::ShowOKDialogText(CVariant{6}, CVariant{477});

Another interactive part in a class I thought that it would be a headless component. Am I wrong? The class name has "GUI" in it. Maybe I'm wrong. Hmm.